Murloc amanyaz bjs2 40538

Amanyaz opened the pull request:
I did the assignment with User-service without tests. I will do tests

InnMish committed:
refactor

InnMish committed:
deleted repository method

mrE100 committed:
subscription service with test of exception

mrE100 committed:
subscription service WITH NO TEST

mrE100 committed:
subscription service WITH NO TEST

mrE100 committed:
SOME tests added

mrE100 committed:
added tests

mrE100 committed:
some annotatiions added

mrE100 committed:
annotation component changed to service and controller

mrE100 committed:
userdto changed to SubscriptionUserDto to avoid conflicts

Amanyaz committed:
Я выполнил задание BJS2-40538 без тестов. Задание было связано с событиями.

alexFromAC143937 commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,88 @@
+@Controller
+public class EventController {
+
+    private EventService eventService;
```
private final EventService eventService;
и конструктор не нужен

Amanyaz replied:
Добавил private final EventService eventService, удалил конструктор с аннотацией @Autowired и вместо него добавил аннотацию @RequiredArgsConstructor.

alexFromAC143937 commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,88 @@
+import java.util.List;
+
+
+@Controller
```
Зависит от того, как будет это использовано, но есть подозрение, что должен быть @RestController

Amanyaz replied:
Сейчас, думаю, для моей задачи этого достаточно, и нет необходимости использовать аннотацию @RestController. Если потребуется, я изменю это в будущем. Согласны?

jibulani replied:
да, для 5-го спринта достаточно даже просто как компонент поставить аннотацию

alexFromAC143937 commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    public void validateEvent(EventDto event) {
+
```
Нужно пеинтересоваться у Евгения, но видится, что зжесь может быть проверка event == null илипроставить на вызове @NotNull,
А условия по ДТО уже стоит проверить в Сервисе

Amanyaz replied:
Давайте оставим как есть. После этой перекрёстной проверки наставник проверит и предложит изменения, если это будет необходимо.

alexFromAC143937 commented on src/main/java/school/faang/user_service/dto/SubscriptionUserDto.java:
```diff
@@ -0,0 +1,10 @@
+import lombok.Data;
+
+@Data
+public class SubscriptionUserDto {
```
Логичнее использовать Record

Amanyaz replied:
Этот класс не был создан мной. Он уже был в проекте, когда я его клонировал. Поэтому было бы правильно оставить всё как есть. Согласны?

alexFromAC143937 commented on src/main/java/school/faang/user_service/repository/recommendation/RecommendationRepository.java:
```diff
@@ -24,7 +24,7 @@ INSERT INTO recommendation (author_id, receiver_id, content)
             """)
     @Modifying
-    Recommendation update(long authorId, long receiverId, String content);
+    void update(long authorId, long receiverId, String content);
```
Думаюю, что не стоит менять сигнатуру разработанного репозитория, если нужно свое, логичным видится добавление ( есть вероятность , что в будущих заданиях будет проблему по реализации и прийдется рефакторить ,то, что ты изменил)

Amanyaz replied:
Я ничего не изменял в классе ReccomendationRepository. Он такой, каким я его склонировал.Я ничего не изменял в классе ReccomendationRepository. Он такой, каким я его склонировал.

alexFromAC143937 commented on src/main/java/school/faang/user_service/repository/recommendation/RecommendationRequestRepository.java:
```diff
@@ -18,11 +18,4 @@ public interface RecommendationRequestRepository extends CrudRepository<Recommen
             LIMIT 1
             """)
     Optional<RecommendationRequest> findLatestPendingRequest(long requesterId, long receiverId);
-
```
Помент по изменению репозитория оставил выше

Amanyaz replied:
Я ничего не изменял в классе ReccomendationRepository. Он такой, каким я его склонировал.

alexFromAC143937 commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,67 @@
+
+    private List<SubscriptionUserDto> filterUsers(Stream<User> users, SubscriptionUserFilterDto filters) {
+
+        userFilters.stream()
```
При написании теста нужно обратить внимание, есть подозрение, что вернет не то, что ожидается, так как 
return userMapper.toDto(users.toList());
при этом
public void apply(Stream<User> users, SubscriptionUserFilterDto subscriptionUserFilterDto)
users - т.е., остается входным стримом без применения фильтра.

Amanyaz replied:
Я не писал классы Subscription и SubscriptionDto, а также их методы. Они такие, какими я их склонировал. Я работал только над частью, связанной с Event.

alexFromAC143937 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,100 @@
+    private final SkillRepository skillRepository;
+    private final EventMapper eventMapper;
+
+    @Autowired
```
Это лишнее, 
    private final EventRepository eventRepository;
    private final SkillRepository skillRepository;
    private final EventMapper eventMapper;
    по такой сигнатуре Спринг сам внедрит зависимости.

Amanyaz replied:
Добавил аннотацию @RequiredArgsConstructor, удалил конструктор и аннотацию @Autowired.Добавил аннотацию @RequiredArgsConstructor, удалил конструктор и аннотацию @Autowired.

alexFromAC143937 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,100 @@
+        List<Event> events = eventRepository.findAll();
+
+
+        return events.stream()
```
Видится логичным иметь реализацию, как у фильтра, тогдаподдерживается расширяемость и не нужно .filter для каждого сравнения писать.

Amanyaz replied:
Сделано.

alexFromAC143937 commented on src/main/java/school/faang/user_service/service/filters/AboutPatternFilter.java:
```diff
@@ -0,0 +1,18 @@
+        return filters.getAboutPattern() != null;
+    }
+
+    @Override
```
Писал выше, не совсем понял логику, есть предположение, что не будет работать, как ожидается.

Amanyaz replied:
Я не создавал этот класс, он уже был в проекте, когда я его клонировал.

alexFromAC143937 commented on src/main/java/school/faang/user_service/service/filters/PageSizeFilter.java:
```diff
@@ -0,0 +1,18 @@
+public class PageSizeFilter implements UserFilter {
+    @Override
+    public boolean isApplicable(SubscriptionUserFilterDto subscriptionUserFilterDto) {
+        return subscriptionUserFilterDto.getPageSize() != 0;
```
getPageSize - может прийти отрицательным, предлагаю поправить условие

Amanyaz replied:
Я не создавал этот класс, он уже был в проекте, когда я его клонировал.

alexFromAC143937 commented on src/main/java/school/faang/user_service/service/filters/PageFilter.java:
```diff
@@ -0,0 +1,18 @@
+public class PageFilter implements UserFilter {
+    @Override
+    public boolean isApplicable(SubscriptionUserFilterDto filters) {
+        return filters.getPage() != 0 && filters.getPageSize() != 0;
```
getPage  || getPageSize - может прийти отрицательным, предлагаю поправить условие

Amanyaz replied:
Я не создавал этот класс, он уже был в проекте, когда я его клонировал.

Amanyaz committed:
После изменений, требуемых при перекрёстной проверке:

1.Создан отдельный метод filterEvent() для класса EventService.
2. Заменены некоторые конструкторы вместе с их аннотацией @Autowired на аннотацию @RequiredArgsConstructor.  (WITHOUT TEST)

jibulani commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,84 @@
+
+    private final EventService eventService;
+
+       public EventDto create(EventDto event) {
```
отступ уехал

Amanyaz replied:
Done

jibulani commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,84 @@
+    public void validateEvent(EventDto event) {
+
+        if (event == null) {
+            throw new IllegalArgumentException("The event cannot be null.");
```
в рамках задачи нужно кастомное исключение кидать:

> В пакете exception для всех исключений в этой задаче создайте свой класс DataValidationException, если его еще нет.

Amanyaz replied:
Done

jibulani commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,84 @@
+
+    }
+
+    public void validateEventUpdate(EventDto event) {
```
зачем отдельный метод, для валидации EventDto должно быть достаточно метода validateEvent

Amanyaz replied:
done

jibulani commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,93 @@
+        return eventMapper.toDto(savedEvent);
+    }
+
+    private boolean hasRequiredSkills(EventDto event) {
```
обычно принято, чтобы приватные методы располагались после всех публичных, хотя и такой вариант, как у тебя тоже встречается
давай все-таки приватные после всех публичных располагать для единообразия всего проекта

Amanyaz replied:
Done

jibulani commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,93 @@
+
+    public EventDto create(EventDto eventDto) {
+        if (!hasRequiredSkills(eventDto)) {
+            throw new IllegalArgumentException("User does not have the required skills to conduct this event");
```
тоже другая ошибка должна быть

Amanyaz replied:
done

jibulani commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,93 @@
+        if (foundEvent.isPresent()) {
+            Event gotEvent = foundEvent.get();
+            return eventMapper.toDto(gotEvent);
+        } else throw new IllegalArgumentException("This event is not exist");
```
можно не писать else, а просто на следующей строке выкинуть исключение, код туда попадет только если не попадет в if - так будет лаконичнее смотреться

Amanyaz replied:
done

jibulani commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,93 @@
+
+    public List<EventDto> filterEvents(List<Event> events, EventFilterDto filter) {
+        return events.stream()
+                .filter(event -> filter.getTitle() == null || event.getTitle().contains(filter.getTitle()))
```
в целом возможный вариант, но неудобный для расширения фильтров
посмотри, как делал фильтры Влад в одном из уроков, по возможности сделай так же

Amanyaz replied:
done

jibulani commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    public EventDto updateEvent(EventDto event) {
+        if (!hasRequiredSkills(event)) {
```
можно весь блок if вынести в отдельный приватный метод, раз он уже дважды повторяется

Amanyaz replied:
done

jibulani reviewed (changes requested):
в целом все хорошо, молодец!
исправь замечания и не забудь покрыть код тестами

Amanyaz committed:
(WITHOUT TEST) После исправления всех комментариев.

Amanyaz committed:
(WITHOUT TEST) После исправления всех комментариев.

jibulani commented on src/main/java/school/faang/user_service/filter/event/EventFilter.java:
```diff
@@ -0,0 +1,14 @@
+
+    boolean isApplicable (EventFilterDto filter);
+
+  Stream<Event> apply(Stream<Event> events, EventFilterDto filter);
```
отступы разъезжаются

Amanyaz replied:
done

jibulani commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -79,14 +79,17 @@ public List<EventDto> getParticipatedEvents(long userId) {
+        return ownerSkillIds.containsAll(requiredSkillIds);
+    }
+
+    private void hasNotRequiredSkills(EventDto event) throws DataValidationException {
```
по названию звучит так, что должен вернуть boolean, а он выкидывает ошибку
я бы назвал validateRequiredSkills

Amanyaz replied:
done

jibulani commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -24,34 +28,32 @@ public class EventService {
-        return filterEvents(events, filter);
+        Stream<Event> eventsStream = events.stream();
+
+        return eventFilters.stream().filter(eventFilter -> eventFilter.isApplicable(eventFilterDto)).reduce(eventsStream,
```
некритично, но я бы для улучшения наглядности так упорядочил:

```
return eventFilters.stream()
                .filter(eventFilter -> eventFilter.isApplicable(eventFilterDto))
                .reduce(eventsStream,(stream, eventFilter) -> eventFilter.apply(stream, eventFilterDto), (s1, s2) -> s1)
                .map(eventMapper::toDto)
                .toList();
```

Amanyaz replied:
done

jibulani reviewed (changes requested):
молодец, что разобрался с фильтрами!
не забудь покрыть тестами и порешать мерж-конфликт

Amanyaz committed:
Merge branch 'Murloc-master-stream7' into Murloc-Amanyaz-BJS2-40538

Amanyaz committed:
With tests

Amanyaz committed:
Merge remote-tracking branch 'origin/Murloc-Amanyaz-BJS2-40538' into Murloc-Amanyaz-BJS2-40538

jibulani reviewed (changes requested):
все ок, но  упали тесты, надо поправить

Amanyaz committed:
With tests
